<a href="https://colab.research.google.com/github/jainam-shah1/Linear-Algebra/blob/main/Singular_Value_Decomposition" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Singular Value Decomposition
>By Jainam Shah

Installing necessary libraries

In [ ]:
install.packages("pracma", repos="http://R-Forge.R-project.org")

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



In [ ]:
library(pracma)
library(MASS)

In [ ]:

install.packages("wordspace")
library(wordspace)

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



In [ ]:
install.packages("caret")
library(caret)

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



Importing all four dataset into individual dataframes

In [ ]:
df_trainoutput <- read.csv('trainOutput.csv', header = FALSE)
df_traininput <- read.csv('trainInput.csv', header = FALSE)
df_testoutput <- read.csv('testOutput.csv', header = FALSE)
df_testinput <- read.csv('testInput.csv', header = FALSE)
dim(df_traininput)
dim(df_trainoutput)
dim(df_testinput)
dim(df_testoutput)

[1]  256 1707

[1]    1 1707

[1]  256 2007

[1]    1 2007

>We have to concatenate both the traininput and trainoutput dataframe using `rbind` and storing it in a dataframe

> New dimension is 257 rows and 1707 columns.

In [ ]:
conc_input <- rbind(df_traininput, df_trainoutput)
dim(conc_input)


[1]  257 1707

> We have to transpose the `conc_input` dataframe.

In [ ]:
conc_input <- t(conc_input)

###Form a matrix A for each digit, such that each row in A represents an image of that digit. (You will have 10 A’s). 

> Convert the `testinput` dataframe in a matrix form.

In [ ]:
df_testinput <- as.matrix(df_testinput)

In [ ]:
mat_A0 = conc_input[conc_input[,257]==0,1:256]
mat_A1 = conc_input[conc_input[,257]==1,1:256]
mat_A2 = conc_input[conc_input[,257]==2,1:256]
mat_A3 = conc_input[conc_input[,257]==3,1:256]
mat_A4 = conc_input[conc_input[,257]==4,1:256]
mat_A5 = conc_input[conc_input[,257]==5,1:256]
mat_A6 = conc_input[conc_input[,257]==6,1:256]
mat_A7 = conc_input[conc_input[,257]==7,1:256]
mat_A8 = conc_input[conc_input[,257]==8,1:256]
mat_A9 = conc_input[conc_input[,257]==9,1:256]

###Determine the singular value decomposition for each A. (Right singular vectors Vi are an orthogonal basis in the image space of that digit. We will refer to the right singular vectors as “singular images.”) You should get 10 sets of singular images, one for each digit. 

In [ ]:
svd0 = svd(mat_A0)
svd1 = svd(mat_A1)
svd2 = svd(mat_A2)
svd3 = svd(mat_A3)
svd4 = svd(mat_A4)
svd5 = svd(mat_A5)
svd6 = svd(mat_A6)
svd7 = svd(mat_A7)
svd8 = svd(mat_A8)
svd9 = svd(mat_A9)


We can now have the singular images for each digit.

In [ ]:
singular0 = as.matrix(svd0$v[,1:20])
singular1 = as.matrix(svd1$v[,1:20])
singular2 = as.matrix(svd2$v[,1:20])  
singular3 = as.matrix(svd3$v[,1:20])  
singular4 = as.matrix(svd4$v[,1:20])  
singular5 = as.matrix(svd5$v[,1:20])
singular6 = as.matrix(svd6$v[,1:20])
singular7 = as.matrix(svd7$v[,1:20])
singular8 = as.matrix(svd8$v[,1:20])
singular9 = as.matrix(svd9$v[,1:20])

###Express test images as a linear combination of the first k=20 singular images of each digit. (This is a least square problem of the form Ax=b). 

In [ ]:
lincom0 = inv(t(singular0) %*% singular0) %*% t(singular0) %*% df_testinput
lincom1 = inv(t(singular1) %*% singular1) %*% t(singular1) %*% df_testinput
lincom2 = inv(t(singular2) %*% singular2) %*% t(singular2) %*% df_testinput
lincom3 = inv(t(singular3) %*% singular3) %*% t(singular3) %*% df_testinput
lincom4 = inv(t(singular4) %*% singular4) %*% t(singular4) %*% df_testinput
lincom5 = inv(t(singular5) %*% singular5) %*% t(singular5) %*% df_testinput
lincom6 = inv(t(singular6) %*% singular6) %*% t(singular6) %*% df_testinput
lincom7 = inv(t(singular7) %*% singular7) %*% t(singular7) %*% df_testinput
lincom8 = inv(t(singular8) %*% singular8) %*% t(singular8) %*% df_testinput
lincom9 = inv(t(singular9) %*% singular9) %*% t(singular9) %*% df_testinput

###Compute the distance between test images and their least square approximations. 

In [ ]:
d0 = t(df_testinput) - t(singular0 %*% lc0)
d1 = t(df_testinput) - t(singular1 %*% lc1)
d2 = t(df_testinput) - t(singular2 %*% lc2)
d3 = t(df_testinput) - t(singular3 %*% lc3)
d4 = t(df_testinput) - t(singular4 %*% lc4)
d5 = t(df_testinput) - t(singular5 %*% lc5)
d6 = t(df_testinput) - t(singular6 %*% lc6)
d7 = t(df_testinput) - t(singular7 %*% lc7)
d8 = t(df_testinput) - t(singular8 %*% lc8)
d9 = t(df_testinput) - t(singular9 %*% lc9)


We have to now find the error and also install the library wordspace

In [ ]:
e0 = rowNorms(as.matrix(d0), method = "euclidean", p = 2)
e1 = rowNorms(as.matrix(d1), method = "euclidean", p = 2)
e2 = rowNorms(as.matrix(d2), method = "euclidean", p = 2)
e3 = rowNorms(as.matrix(d3), method = "euclidean", p = 2)
e4 = rowNorms(as.matrix(d4), method = "euclidean", p = 2)
e5 = rowNorms(as.matrix(d5), method = "euclidean", p = 2)
e6 = rowNorms(as.matrix(d6), method = "euclidean", p = 2)
e7 = rowNorms(as.matrix(d7), method = "euclidean", p = 2)
e8 = rowNorms(as.matrix(d8), method = "euclidean", p = 2)
e9 = rowNorms(as.matrix(d9), method = "euclidean", p = 2)


Finding the error

In [ ]:
error_df = as.data.frame(cbind(e0, e1, e2, e3, e4, e5, e6, e7, e8, e9))
head(error_df,20)

e0        e1        e2       e3        e4        e5        e6       
V1   9.450711 10.895408 9.569496  8.691607  7.284899  9.337940  9.545622
V2   7.987032 11.496614 9.112643 10.416280  9.436814 10.232620  5.514822
V3   8.947269 12.148701 8.665643  4.168662 11.736596  9.105364 10.195266
V4   7.747875 11.188172 9.514209  9.945517  9.446872 10.272118  7.928857
V5   8.588480 11.414212 9.804297 11.469535 10.496134 11.174644  5.815101
V6   3.581011 11.474235 7.884316 10.103205 11.312008  8.516164  8.556714
V7   3.500403 11.657214 7.344438  8.418468 10.684792  8.568493  8.173763
V8   3.903384 11.655254 8.592899  8.474108 10.690676  8.133569  7.516063
V9   8.222032 11.905395 9.130111 10.182423  9.420101  9.318434  3.059844
V10  8.134053  8.722042 8.144375  7.808162  6.548223  8.571951  8.037554
V11  8.524713 10.415999 9.431173  8.780232  9.527240  7.579814  3.271304
V12  9.977934 11.043389 4.822555  9.706148 10.000929 11.067303 10.177089
V13  7.309545 10.755568 6.696660  6.647040 12.107406  8.704424 10.310586
V14 10.282615 10.435126 9.127514  9.958588  7.741412  9.470704  9.509110
V15  7.010918 11.516318 9.520462 10.585501 10.154173  8.473217  9.781461
V16  8.548656  9.702626 9.072556  6.029274  8.665472  7.362178  9.307555
V17  8.434998  2.239679 8.295990  8.681904  6.147263  8.603323  6.606260
V18  9.529732 10.646817 9.014435 11.013111  8.925620  9.635278  8.530432
V19  9.874296 11.958149 9.727099  9.946570  8.327298 10.505131 11.068625
V20  8.239734  9.562780 9.458614  7.978047  8.941878  7.145134  3.992378
    e7        e8        e9       
V1   7.114044  8.236318  3.401388
V2  11.448302 10.653984 10.843435
V3   9.879520  7.773044  9.382345
V4  11.312533 10.776754 11.051277
V5  11.437068 10.891941 11.893676
V6  10.911723 10.670079 11.776834
V7   9.239416  8.801773  9.217634
V8  10.120198  8.419270  9.948246
V9  11.436944  8.937615 11.109879
V10  5.834232  6.772362  3.436013
V11 11.205966  8.903856 10.502163
V12 10.337096 10.681799 11.141079
V13 11.512148 10.543030 11.666985
V14  7.786888  8.451192  7.944845
V15  9.988218 10.785290 10.834752
V16  8.525110  7.221609  8.281699
V17  8.018654  5.950564  7.319382
V18 11.186678 10.421931 10.910942
V19  7.662169  7.440991  5.054926
V20 11.957092  9.116176 11.441232

###Classify each test image to be the digit corresponding to the smallest residual.  

In [ ]:
prediction = apply(error_df, 1, which.min) - 1


In [ ]:
classification_rate = sum(prediction == df_testoutput) / ncol(df_testoutput)
classification_rate*100

[1] 93.9711

###Calculate the overall correct classification rate, as well as correct classification rate for each digit in a confusion matrix. 

In [ ]:
confusionMatrix(factor(df_testoutput), factor(prediction))

Warning message in xtfrm.data.frame(x):
“cannot xtfrm data frames”


Confusion Matrix and Statistics

          Reference
Prediction   0   1   2   3   4   5   6   7   8   9
         0 355   0   2   0   1   0   0   0   0   1
         1   0 259   0   0   3   0   2   0   0   0
         2   8   1 179   2   3   0   0   1   4   0
         3   1   0   4 148   1   8   0   1   2   1
         4   1   1   0   0 187   1   1   3   0   6
         5   8   1   2   4   0 140   0   0   2   3
         6   2   0   0   0   2   2 164   0   0   0
         7   0   1   1   0   3   0   0 139   0   3
         8   4   0   2   6   0   1   0   0 149   4
         9   0   1   0   1   4   0   0   4   1 166

Overall Statistics
                                          
               Accuracy : 0.9397          
                 95% CI : (0.9284, 0.9497)
    No Information Rate : 0.1888          
    P-Value [Acc > NIR] : < 2.2e-16       
                                          
                  Kappa : 0.9322          
                                          
 Mcnemar's Test P-Valu

> As per confusion matrix the **accuracy** is *93.97%*